### ISSUE TITLE
[Bithack] Handle xor of double with 4602678819172646912

### ISSUE THREAD

**Issue by ZuseZ4** [EnzymeAD/Enzyme#689]
To test Enzyme for higher orders and improve our docs I tried to follow the[ jax cookbook](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html). It fails however pretty early due to Eigen introducing some integer operation which Enzyme can't handle.
https://fwd.gymni.ch/UQclGh

```
...
 constantinst[  %217 = fadd <2 x double> %215, %216] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %218 = extractelement <2 x double> %217, i32 0] = 0 val:0 type: {[-1]:Float@double}
 constantinst[  %219 = fneg double %218] = 0 val:0 type: {}
 constantinst[  call void @llvm.lifetime.end.p0i8(i64 32, i8* nonnull %6) #14] = 1 val:1 type: {}
 constantinst[  ret double %219] = 1 val:1 type: {}
cannot handle unknown binary operator:   %174 = and <2 x i64> %173, %171
```

**Comment by wsmoses** [comment]
with loose type analysis:
`` cannot handle unknown binary operator:   %168 = or <2 x i64> %167, <i64 4602678819172646912, i64 4602678819172646912> ``

**Comment by wsmoses** [comment]
Inversion issue: https://fwd.gymni.ch/kGTcQM

**Comment by ZuseZ4** [comment]
Another unknown binary operator: `sub`
https://fwd.gymni.ch/Lwx9Gw

**Comment by wsmoses** [comment]
The last error here is that you need to run with at least O0, enable TBAA, or enable loose type analysis.

**Comment by ZuseZ4** [comment]
constantinst[  br label %246] = 1 val:1 type: {}
 constantinst[  ret void] = 1 val:1 type: {}
cannot handle unknown binary operator:   %109 = ashr exact i64 %108, 4

**Comment by ZuseZ4** [comment]
```.ll
121334:updating analysis of val:   %155 = phi %"struct.std::pair.265"* [ %94, %90 ], [ %222, %218 ] current: {} new {} from   %155 = phi %"struct.std::pair.265"* [ %94, %90 ], [ %222, %218 ] Changed=0 legal=1
121379:updating analysis of val:   %155 = phi %"struct.std::pair.265"* [ %94, %90 ], [ %222, %218 ] current: {} new {} from   %171 = icmp eq %"struct.std::pair.265"* %155, %154 Changed=0 legal=1
121488:updating analysis of val:   %212 = phi %"struct.std::pair.265"* [ %193, %182 ], [ %208, %210 ] current: {} new {[-1]:Pointer} from   %212 = phi %"struct.std::pair.265"* [ %193, %182 ], [ %208, %210 ] Changed=1 legal=1
121498:updating analysis of val:   %212 = phi %"struct.std::pair.265"* [ %193, %182 ], [ %208, %210 ] current: {[-1]:Pointer} new {} from   %219 = phi %"struct.std::pair.265"* [ %212, %216 ], [ %155, %172 ] Changed=0 legal=1
121500:updating analysis of val:   %219 = phi %"struct.std::pair.265"* [ %212, %216 ], [ %155, %172 ] current: {} new {} from   %219 = phi %"struct.std::pair.265"* [ %212, %216 ], [ %155, %172 ] Changed=0 legal=1
121509:updating analysis of val:   %219 = phi %"struct.std::pair.265"* [ %212, %216 ], [ %155, %172 ] current: {} new {} from   %222 = getelementptr inbounds %"struct.std::pair.265", %"struct.std::pair.265"* %219, i64 1 Changed=0 legal=1
121511:updating analysis of val:   %219 = phi %"struct.std::pair.265"* [ %212, %216 ], [ %155, %172 ] current: {} new {} from   %222 = getelementptr inbounds %"struct.std::pair.265", %"struct.std::pair.265"* %219, i64 1 Changed=0 legal=1
121789:updating analysis of val:   %155 = phi %"struct.std::pair.265"* [ %94, %90 ], [ %222, %218 ] current: {} new {} from   %155 = phi %"struct.std::pair.265"* [ %94, %90 ], [ %222, %218 ] Changed=0 legal=1
121967:updating analysis of val:   %155 = phi %"struct.std::pair.265"* [ %94, %90 ], [ %222, %218 ] current: {} new {[-1]:Pointer} from   %171 = icmp eq %"struct.std::pair.265"* %155, %154 Changed=1 legal=1
121985:updating analysis of val:   %212 = phi %"struct.std::pair.265"* [ %193, %182 ], [ %208, %210 ] current: {[-1]:Pointer} new {[-1]:Pointer} from   %212 = phi %"struct.std::pair.265"* [ %193, %182 ], [ %208, %210 ] Changed=0 legal=1
122031:updating analysis of val:   %155 = phi %"struct.std::pair.265"* [ %94, %90 ], [ %222, %218 ] current: {[-1]:Pointer} new {} from   %155 = phi %"struct.std::pair.265"* [ %94, %90 ], [ %222, %218 ] Changed=0 legal=1
122032:updating analysis of val:   %212 = phi %"struct.std::pair.265"* [ %193, %182 ], [ %208, %210 ] current: {[-1]:Pointer} new {} from   %219 = phi %"struct.std::pair.265"* [ %212, %216 ], [ %155, %172 ] Changed=0 legal=1
122034:updating analysis of val:   %219 = phi %"struct.std::pair.265"* [ %212, %216 ], [ %155, %172 ] current: {} new {} from   %219 = phi %"struct.std::pair.265"* [ %212, %216 ], [ %155, %172 ] Changed=0 legal=1
122049:updating analysis of val:   %219 = phi %"struct.std::pair.265"* [ %212, %216 ], [ %155, %172 ] current: {} new {} from   %222 = getelementptr inbounds %"struct.std::pair.265", %"struct.std::pair.265"* %219, i64 1 Changed=0 legal=1
122051:updating analysis of val:   %219 = phi %"struct.std::pair.265"* [ %212, %216 ], [ %155, %172 ] current: {} new {} from   %222 = getelementptr inbounds %"struct.std::pair.265", %"struct.std::pair.265"* %219, i64 1 Changed=0 legal=1
```
For some reason Enzyme fails to get the type of 219 and as a consequence of the ashr line.
```.ll
constantinst[  %155 = phi %"struct.std::pair.265"* [ %94, %90 ], [ %222, %218 ]] = 1 val:0 type: {[-1]:Pointer}
constantinst[  %212 = phi %"struct.std::pair.265"* [ %193, %182 ], [ %208, %210 ]] = 1 val:0 type: {[-1]:Pointer}
constantinst[  %219 = phi %"struct.std::pair.265"* [ %212, %216 ], [ %155, %172 ]] = 0 val:0 type: {}
```

**Comment by wsmoses** [comment]
@ZuseZ4 at this point this bug I believe comes down to supporting differentiating through the equivalent of the bithack of (double value) or with (that bitstring). We have a few examples of bithacks in the code already.

If this is an important issue for you presently, the path forward for it would to basically make a minimal function like the following as a test, determine from the bitwise operation itself the semantic meaning of the operation, and add a corresponding derivative.
```
Double f(double x) {
  Return x | MAGIC_NUMBER;
}
```

**Comment by ZuseZ4** [comment]
I think this bithack here was only active due to an AA/TA bug in the first place, which was solved I think. So I'll close for now (and eventually reopen if I was wrong)

**Comment by wsmoses** [comment]
I think this still may be useful to handle as a bithack, so I’m going to leave it open to preserve.

**Comment by wsmoses** [comment]
This also still persists on the compiler explorer code at the top.

**Comment by ZuseZ4** [comment]
Indeed, I guess then I just got confused between this and another unknown binary operator assertion. It would then be good to have this to be able to move over the jax examples to Enzyme for documentation purpose. Will try to go back once I have the OpenLB bugs.
